File: packages/python/goldenmatch/goldenmatch/plugins/builtin/business.py
"""Business-shaped predefined plugins.

system_of_record, lifecycle_stage, freshness_with_max_age. Each
satisfies ``GoldenStrategyPlugin`` from ``goldenmatch.plugins.base``.

Spec: ``docs/superpowers/specs/2026-05-22-predefined-merge-plugins-design.md``
"""
from __future__ import annotations

import math
import re
from datetime import UTC, datetime, timedelta
from typing import Any

# Default order for `lifecycle_stage` (low -> high). Lowercased for
# case-insensitive matching.
DEFAULT_LIFECYCLE_ORDER = [
    "subscriber",
    "lead",
    "marketing_qualified_lead",
    "mql",
    "sales_qualified_lead",
    "sql",
    "opportunity",
    "customer",
    "evangelist",
]


class SystemOfRecordStrategy:
    """Pick value from authoritative source per `rule_kwargs.source_priority`.

    Semantically equivalent to the built-in `source_priority` strategy
    but with explicit "system of record" naming so the YAML config
    carries operator intent:

    .. code-block:: yaml

       golden_rules:
         field_rules:
           account_status:
             strategy: "custom:system_of_record"
             source_priority: ["salesforce", "hubspot", "netsuite"]

    Falls back to first non-null when no priority source has a value.
    """

    name = "system_of_record"

    def merge(
        self,
        values: list,
        *,
        sources: list[str] | None = None,
        rule_kwargs: dict | None = None,
        **_: Any,
    ) -> Any:
        non_null = [(i, v) for i, v in enumerate(values) if v is not None]
        if not non_null:
            return (None, 0.0)
        priority = (rule_kwargs or {}).get("source_priority") or []
        if sources is None or not priority:
            # No priority config -> first non-null with low confidence
            # (the YAML intent says "system of record" but no SoR is
            # configured, so this is a fallback).
            return (non_null[0][1], 0.5, non_null[0][0])
        for src in priority:
            for i, (s, v) in enumerate(zip(sources, values)):
                if s == src and v is not None:
                    # Confidence decays with priority rank
                    rank = priority.index(src)
                    conf = max(0.5, 1.0 - rank * 0.1)
                    return (v, conf, i)
        # No source in priority list had a value -> first non-null.
        return (non_null[0][1], 0.4, non_null[0][0])


class LifecycleStageStrategy:
    """Pick the most-advanced lifecycle stage from the cluster.

    Default order (low -> high): subscriber, lead, mql, sql,
    opportunity, customer, evangelist. Override via
    `rule_kwargs.lifecycle_order` (list of stage names; comparison
    is lowercase-insensitive).

    Unknown stages are ranked below the lowest known stage
    (effectively ignored when ANY known stage is present).

    Confidence: 1.0 when a unique maximum stage exists; 0.7 on ties.
    """

    name = "lifecycle_stage"

    def merge(
        self,
        values: list,
        *,
        rule_kwargs: dict | None = None,
        **_: Any,
    ) -> Any:
        non_null = [(i, v) for i, v in enumerate(values) if v is not None]
        if not non_null:
            return (None, 0.0)
        order = (rule_kwargs or {}).get("lifecycle_order") or DEFAULT_LIFECYCLE_ORDER
        order_map = {str(stage).strip().lower(): idx for idx, stage in enumerate(order)}
        # Unknown values: rank = -1 (below all known stages).
        ranked = [
            (order_map.get(str(v).strip().lower(), -1), i, v)
            for i, v in non_null
        ]
        max_rank = max(r for r, _, _ in ranked)
        tied = [(i, v) for r, i, v in ranked if r == max_rank]
        conf = 1.0 if len(tied) == 1 else 0.7
        return (tied[0][1], conf, tied[0][0])


def _parse_date(value: Any) -> datetime | None:
    """Best-effort datetime coercion. Handles common ISO + epoch."""
    if value is None:
        return None
    if isinstance(value, datetime):
        return value if value.tzinfo else value.replace(tzinfo=UTC)
    if isinstance(value, (int, float)) and not isinstance(value, bool):
        try:
            return datetime.fromtimestamp(float(value), tz=UTC)
        except (OSError, OverflowError, ValueError):
            return None
    s = str(value).strip()
    if not s:
        return None
    # Try ISO 8601 variants. Python 3.11+ handles most.
    try:
        dt = datetime.fromisoformat(s.replace("Z", "+00:00"))
        return dt if dt.tzinfo else dt.replace(tzinfo=UTC)
    except ValueError:
        pass
    # Common fallback patterns.
    for fmt in ("%Y-%m-%d", "%Y/%m/%d", "%m/%d/%Y", "%d/%m/%Y", "%Y-%m-%d %H:%M:%S"):
        try:
            dt = datetime.strptime(s, fmt)
            return dt.replace(tzinfo=UTC)
        except ValueError:
            continue
    return None


class FreshnessWithMaxAgeStrategy:
    """Like `most_recent` but emits NULL if no value is fresh enough.

    `rule_kwargs.max_age_days` (default 365) defines the cutoff.
    Values whose accompanying date is older than cutoff are dropped.
    If no values remain after filtering, emits (None, 0.0).

    Use when compliance / data-quality requires "no stale data" --
    e.g. KYC address must be < 90 days old, else missing-data
    follow-up triggers.

    Requires `dates` kwarg (parallel to values). Without dates,
    behaves as if every value is stale (emits None).
    """

    name = "freshness_with_max_age"

    def merge(
        self,
        values: list,
        *,
        dates: list | None = None,
        rule_kwargs: dict | None = None,
        **_: Any,
    ) -> Any:
        if dates is None:
            return (None, 0.0)
        max_age_days = float((rule_kwargs or {}).get("max_age_days", 365))
        now = datetime.now(tz=UTC)
        cutoff = now - timedelta(days=max_age_days)
        candidates: list[tuple[int, datetime, Any]] = []
        for i, (d, v) in enumerate(zip(dates, values)):
            if v is None:
                continue
            parsed = _parse_date(d)
            if parsed is None or parsed < cutoff:
                continue
            candidates.append((i, parsed, v))
        if not candidates:
            return (None, 0.0)
        # Sort newest-first.
        candidates.sort(key=lambda x: x[1], reverse=True)
        top_dt = candidates[0][1]
        tied = [(i, v) for i, dt, v in candidates if dt == top_dt]
        conf = 1.0 if len(tied) == 1 else 0.7
        return (tied[0][1], conf, tied[0][0])


class EnumCanonicalStrategy:
    """Map known aliases to a canonical enum value, then pick the mode.

    `rule_kwargs.alias_map` maps any alias (case-insensitive,
    trimmed) -> canonical value. For example::

       alias_map:
         "USA": "US"
         "United States": "US"
         "U.S.": "US"
         "U.S.A.": "US"

    Lookup is case-insensitive on KEYS but the VALUE is returned
    verbatim (preserves intentional casing of the canonical form).

    Values not in the alias_map pass through unchanged (no
    modification). Then the strategy picks the mode of the resulting
    set.

    Confidence: `count(winning_value) / count(non_null)`.
    """

    name = "enum_canonical"

    def merge(
        self,
        values: list,
        *,
        rule_kwargs: dict | None = None,
        **_: Any,
    ) -> Any:
        from collections import Counter

        non_null = [(i, v) for i, v in enumerate(values) if v is not None]
        if not non_null:
            return (None, 0.0)
        raw_map: dict = (rule_kwargs or {}).get("alias_map") or {}
        # Build a case-insensitive lookup -- key.lower() -> canonical.
        alias_map = {str(k).strip().lower(): canonical for k, canonical in raw_map.items()}

        def _canonicalize(v: Any) -> Any:
            key = str(v).strip().lower()
            return alias_map.get(key, v)

        normalized = [(i, _canonicalize(v)) for i, v in non_null]
        counts = Counter(c for _, c in normalized)
        winner, count = counts.most_common(1)[0]
        first_idx = next(i for i, c in normalized if c == winner)
        conf = count / len(non_null)
        return (winner, conf, first_idx)


class RegexValidatedStrategy:
    """Only accept values matching ``rule_kwargs.pattern``.

    Useful when source data is dirty and you want to filter at the
    consolidation step (e.g. only accept email-shaped values into the
    `email` field, only accept SSN-shaped values into `ssn`).

    Behavior:
    - Filter values to only those matching the regex (`re.fullmatch`).
    - If at least one matches: pick the most-common one (mode).
    - If none match: fall back per ``rule_kwargs.fallback``:
        - "first_non_null" (default): pick the first non-null
          unmatched value with low confidence (0.3)
        - "null": emit (None, 0.0)

    Missing ``rule_kwargs.pattern`` -> behaves like `first_non_null`.
    """

    name = "regex_validated"

    def merge(
        self,
        values: list,
        *,
        rule_kwargs: dict | None = None,
        **_: Any,
    ) -> Any:
        from collections import Counter

        non_null = [(i, v) for i, v in enumerate(values) if v is not None]
        if not non_null:
            return (None, 0.0)
        pattern = (rule_kwargs or {}).get("pattern")
        fallback = (rule_kwargs or {}).get("fallback", "first_non_null")
        if not pattern:
            # No pattern configured -> first non-null.
            return (non_null[0][1], 0.5, non_null[0][0])
        try:
            compiled = re.compile(pattern)
        except re.error:
            return (non_null[0][1], 0.3, non_null[0][0])
        matched = [(i, v) for i, v in non_null if compiled.fullmatch(str(v))]
        if matched:
            counts = Counter(v for _, v in matched)
            winner, count = counts.most_common(1)[0]
            first_idx = next(i for i, v in matched if v == winner)
            conf = count / len(matched)
            return (winner, conf, first_idx)
        # No matches.
        if fallback == "null":
            return (None, 0.0)
        return (non_null[0][1], 0.3, non_null[0][0])


class WeightedByRecencyStrategy:
    """Pick the value with the highest exponential-decayed recency weight.

    Each value's weight = exp(-age_days / half_life_days). The value
    with the maximum weight is returned. When multiple values share
    the same date, the first-index wins.

    `rule_kwargs.half_life_days` (default 30) controls decay rate:
    after `half_life_days` days, weight halves. A value 6 months
    older has weight ~0.001 of a fresh value (effectively ignored).

    Requires `dates` kwarg. Values without dates are dropped.
    Confidence: 1.0 when unique winner; 0.7 on date ties.

    Compare to `most_recent` (sharp cutoff at newest) and
    `freshness_with_max_age` (binary in/out). This strategy is a soft
    "newer-but-not-only-newest" picker.
    """

    name = "weighted_by_recency"

    def merge(
        self,
        values: list,
        *,
        dates: list | None = None,
        rule_kwargs: dict | None = None,
        **_: Any,
    ) -> Any:
        if dates is None:
            return (None, 0.0)
        half_life_days = float((rule_kwargs or {}).get("half_life_days", 30))
        if half_life_days <= 0:
            half_life_days = 30.0
        now = datetime.now(tz=UTC)
        scored: list[tuple[int, datetime, float, Any]] = []
        for i, (d, v) in enumerate(zip(dates, values)):
            if v is None:
                continue
            parsed = _parse_date(d)
            if parsed is None:
                continue
            age_days = (now - parsed).total_seconds() / 86400.0
            weight = math.exp(-age_days / half_life_days)
            scored.append((i, parsed, weight, v))
        if not scored:
            return (None, 0.0)
        scored.sort(key=lambda x: x[2], reverse=True)
        top_weight = scored[0][2]
        tied = [(i, v) for i, _, w, v in scored if w == top_weight]
        conf = 1.0 if len(tied) == 1 else 0.7
        return (tied[0][1], conf, tied[0][0])

File: packages/typescript/goldenmatch/src/core/plugins/builtin/business.ts
/**
 * Business-shaped golden strategies (Phase 5 Part 3 of N -- closes
 * the next 6 of 22 plugins from goldenmatch issue #208).
 *
 * Includes the date-handling pair (`freshness_with_max_age` +
 * `weighted_by_recency`) -- their parity-test fixtures use dates
 * with stable relative ordering so the winner is deterministic
 * regardless of when the test runs.
 */

import type {
  GoldenStrategyMergeOpts,
  GoldenStrategyPlugin,
  GoldenStrategyResult,
} from "../base.js";

const DEFAULT_LIFECYCLE_ORDER: ReadonlyArray<string> = [
  "subscriber",
  "lead",
  "marketing_qualified_lead",
  "mql",
  "sales_qualified_lead",
  "sql",
  "opportunity",
  "customer",
  "evangelist",
];

function pyStr(value: unknown): string {
  if (value === null || value === undefined) return "";
  if (typeof value === "boolean") return value ? "True" : "False";
  return String(value);
}

function mostCommon<T>(items: ReadonlyArray<T>): { key: T; count: number } {
  const counts = new Map<T, number>();
  for (const item of items) counts.set(item, (counts.get(item) ?? 0) + 1);
  let bestKey: T | undefined;
  let bestCount = -1;
  for (const [key, count] of counts) {
    if (count > bestCount) {
      bestCount = count;
      bestKey = key;
    }
  }
  return { key: bestKey as T, count: bestCount };
}

function parseDate(value: unknown): Date | null {
  if (value === null || value === undefined) return null;
  if (value instanceof Date) {
    return Number.isFinite(value.getTime()) ? value : null;
  }
  if (typeof value === "number" && Number.isFinite(value)) {
    // Python `datetime.fromtimestamp(float)` treats as seconds.
    return new Date(value * 1000);
  }
  if (typeof value === "boolean") return null;
  const s = String(value).trim();
  if (!s) return null;
  // Python's `datetime.fromisoformat` accepts most ISO 8601;
  // JS `Date.parse` handles the standard forms used in fixtures
  // ("2024-01-15", "2024-01-15T12:00:00Z", "2024-01-15T12:00:00+00:00").
  const ms = Date.parse(s);
  if (Number.isFinite(ms)) return new Date(ms);
  // Python fallbacks (`%Y/%m/%d`, `%m/%d/%Y`, etc.) -- handle the
  // unambiguous slash form. The ambiguous d/m/Y vs m/d/Y case is
  // not exercised in fixtures (we always use ISO).
  const slashMatch = s.match(/^(\d{4})\/(\d{2})\/(\d{2})$/);
  if (slashMatch) {
    return new Date(`${slashMatch[1]}-${slashMatch[2]}-${slashMatch[3]}T00:00:00Z`);
  }
  return null;
}

export class SystemOfRecordStrategy implements GoldenStrategyPlugin {
  readonly name = "system_of_record";

  merge(
    values: ReadonlyArray<unknown>,
    opts: GoldenStrategyMergeOpts = {},
  ): GoldenStrategyResult {
    const nonNull: Array<{ idx: number; raw: unknown }> = [];
    for (let i = 0; i < values.length; i++) {
      if (values[i] !== null && values[i] !== undefined) {
        nonNull.push({ idx: i, raw: values[i] });
      }
    }
    if (nonNull.length === 0) return [null, 0.0] as const;
    const sources = opts.sources;
    const priority =
      (opts.ruleKwargs?.["source_priority"] as readonly string[] | undefined) ??
      [];
    if (!sources || priority.length === 0) {
      return [nonNull[0]!.raw, 0.5, nonNull[0]!.idx] as const;
    }
    for (let rank = 0; rank < priority.length; rank++) {
      const targetSrc = priority[rank];
      for (let i = 0; i < sources.length; i++) {
        if (sources[i] === targetSrc && values[i] !== null && values[i] !== undefined) {
          const conf = Math.max(0.5, 1.0 - rank * 0.1);
          return [values[i], conf, i] as const;
        }
      }
    }
    return [nonNull[0]!.raw, 0.4, nonNull[0]!.idx] as const;
  }
}

export class LifecycleStageStrategy implements GoldenStrategyPlugin {
  readonly name = "lifecycle_stage";

  merge(
    values: ReadonlyArray<unknown>,
    opts: GoldenStrategyMergeOpts = {},
  ): GoldenStrategyResult {
    const nonNull: Array<{ idx: number; raw: unknown }> = [];
    for (let i = 0; i < values.length; i++) {
      if (values[i] !== null && values[i] !== undefined) {
        nonNull.push({ idx: i, raw: values[i] });
      }
    }
    if (nonNull.length === 0) return [null, 0.0] as const;
    const order =
      (opts.ruleKwargs?.["lifecycle_order"] as readonly unknown[] | undefined) ??
      DEFAULT_LIFECYCLE_ORDER;
    const orderMap = new Map<string, number>();
    for (let i = 0; i < order.length; i++) {
      orderMap.set(pyStr(order[i]).trim().toLowerCase(), i);
    }
    const ranked = nonNull.map((x) => ({
      idx: x.idx,
      raw: x.raw,
      rank: orderMap.get(pyStr(x.raw).trim().toLowerCase()) ?? -1,
    }));
    let maxRank = ranked[0]!.rank;
    for (let i = 1; i < ranked.length; i++) {
      if (ranked[i]!.rank > maxRank) maxRank = ranked[i]!.rank;
    }
    const tied = ranked.filter((x) => x.rank === maxRank);
    const conf = tied.length === 1 ? 1.0 : 0.7;
    return [tied[0]!.raw, conf, tied[0]!.idx] as const;
  }
}

export class FreshnessWithMaxAgeStrategy implements GoldenStrategyPlugin {
  readonly name = "freshness_with_max_age";

  merge(
    values: ReadonlyArray<unknown>,
    opts: GoldenStrategyMergeOpts = {},
  ): GoldenStrategyResult {
    const dates = opts.dates;
    if (!dates) return [null, 0.0] as const;
    const maxAgeDays =
      Number((opts.ruleKwargs?.["max_age_days"] as number | undefined) ?? 365);
    // Allow tests to pin `now` via ruleKwargs.now_iso for determinism.
    const nowIso = opts.ruleKwargs?.["now_iso"] as string | undefined;
    const now = nowIso ? new Date(nowIso) : new Date();
    const cutoffMs = now.getTime() - maxAgeDays * 86400 * 1000;
    const candidates: Array<{ idx: number; dt: Date; raw: unknown }> = [];
    for (let i = 0; i < values.length; i++) {
      const v = values[i];
      if (v === null || v === undefined) continue;
      const dt = parseDate(dates[i]);
      if (dt === null || dt.getTime() < cutoffMs) continue;
      candidates.push({ idx: i, dt, raw: v });
    }
    if (candidates.length === 0) return [null, 0.0] as const;
    candidates.sort((a, b) => b.dt.getTime() - a.dt.getTime());
    const topMs = candidates[0]!.dt.getTime();
    const tied = candidates.filter((x) => x.dt.getTime() === topMs);
    const conf = tied.length === 1 ? 1.0 : 0.7;
    return [tied[0]!.raw, conf, tied[0]!.idx] as const;
  }
}

export class EnumCanonicalStrategy implements GoldenStrategyPlugin {
  readonly name = "enum_canonical";

  merge(
    values: ReadonlyArray<unknown>,
    opts: GoldenStrategyMergeOpts = {},
  ): GoldenStrategyResult {
    const nonNull: Array<{ idx: number; raw: unknown }> = [];
    for (let i = 0; i < values.length; i++) {
      if (values[i] !== null && values[i] !== undefined) {
        nonNull.push({ idx: i, raw: values[i] });
      }
    }
    if (nonNull.length === 0) return [null, 0.0] as const;
    const rawMap =
      (opts.ruleKwargs?.["alias_map"] as Record<string, unknown> | undefined) ??
      {};
    const aliasMap = new Map<string, unknown>();
    for (const [k, canonical] of Object.entries(rawMap)) {
      aliasMap.set(pyStr(k).trim().toLowerCase(), canonical);
    }
    const canonicalize = (v: unknown): unknown => {
      const key = pyStr(v).trim().toLowerCase();
      const mapped = aliasMap.get(key);
      return mapped !== undefined ? mapped : v;
    };
    const normalized = nonNull.map((x) => ({ idx: x.idx, c: canonicalize(x.raw) }));
    const { key: winner, count } = mostCommon(normalized.map((x) => x.c));
    const firstIdx = normalized.find(
      (x) => x.c === winner || JSON.stringify(x.c) === JSON.stringify(winner),
    )!.idx;
    return [winner, count / nonNull.length, firstIdx] as const;
  }
}

export class RegexValidatedStrategy implements GoldenStrategyPlugin {
  readonly name = "regex_validated";

  merge(
    values: ReadonlyArray<unknown>,
    opts: GoldenStrategyMergeOpts = {},
  ): GoldenStrategyResult {
    const nonNull: Array<{ idx: number; raw: unknown }> = [];
    for (let i = 0; i < values.length; i++) {
      if (values[i] !== null && values[i] !== undefined) {
        nonNull.push({ idx: i, raw: values[i] });
      }
    }
    if (nonNull.length === 0) return [null, 0.0] as const;
    const pattern = opts.ruleKwargs?.["pattern"] as string | undefined;
    const fallback =
      (opts.ruleKwargs?.["fallback"] as string | undefined) ?? "first_non_null";
    if (!pattern) {
      return [nonNull[0]!.raw, 0.5, nonNull[0]!.idx] as const;
    }
    let compiled: RegExp;
    try {
      // Python's `re.fullmatch` -> anchor both ends explicitly.
      compiled = new RegExp(`^(?:${pattern})$`);
    } catch {
      return [nonNull[0]!.raw, 0.3, nonNull[0]!.idx] as const;
    }
    const matched = nonNull.filter((x) => compiled.test(pyStr(x.raw)));
    if (matched.length > 0) {
      const { key: winner, count } = mostCommon(matched.map((x) => x.raw));
      const firstIdx = matched.find(
        (x) => x.raw === winner || JSON.stringify(x.raw) === JSON.stringify(winner),
      )!.idx;
      return [winner, count / matched.length, firstIdx] as const;
    }
    if (fallback === "null") return [null, 0.0] as const;
    return [nonNull[0]!.raw, 0.3, nonNull[0]!.idx] as const;
  }
}

export class WeightedByRecencyStrategy implements GoldenStrategyPlugin {
  readonly name = "weighted_by_recency";

  merge(
    values: ReadonlyArray<unknown>,
    opts: GoldenStrategyMergeOpts = {},
  ): GoldenStrategyResult {
    const dates = opts.dates;
    if (!dates) return [null, 0.0] as const;
    let halfLifeDays = Number(
      (opts.ruleKwargs?.["half_life_days"] as number | undefined) ?? 30,
    );
    if (!(halfLifeDays > 0)) halfLifeDays = 30;
    const nowIso = opts.ruleKwargs?.["now_iso"] as string | undefined;
    const now = nowIso ? new Date(nowIso) : new Date();
    const scored: Array<{ idx: number; dt: Date; weight: number; raw: unknown }> = [];
    for (let i = 0; i < values.length; i++) {
      const v = values[i];
      if (v === null || v === undefined) continue;
      const dt = parseDate(dates[i]);
      if (dt === null) continue;
      const ageDays = (now.getTime() - dt.getTime()) / 86_400_000;
      const weight = Math.exp(-ageDays / halfLifeDays);
      scored.push({ idx: i, dt, weight, raw: v });
    }
    if (scored.length === 0) return [null, 0.0] as const;
    scored.sort((a, b) => b.weight - a.weight);
    const topWeight = scored[0]!.weight;
    const tied = scored.filter((x) => x.weight === topWeight);
    const conf = tied.length === 1 ? 1.0 : 0.7;
    return [tied[0]!.raw, conf, tied[0]!.idx] as const;
  }
}

export const BUSINESS_BUILTINS: readonly GoldenStrategyPlugin[] = [
  new SystemOfRecordStrategy(),
  new LifecycleStageStrategy(),
  new FreshnessWithMaxAgeStrategy(),
  new EnumCanonicalStrategy(),
  new RegexValidatedStrategy(),
  new WeightedByRecencyStrategy(),
] as const;
